module/zfs/spa.c

```diff
@@ -9049,8 +9059,13 @@ spa_async_thread(void *arg)
 	 */
 	if (tasks & SPA_ASYNC_FAULT_VDEV) {
 		spa_vdev_state_enter(spa, SCL_NONE);
-		spa_async_fault_vdev(spa, spa->spa_root_vdev);
+		boolean_t suspend = B_FALSE;
+		spa_async_fault_vdev(spa->spa_root_vdev, &suspend);
 		(void) spa_vdev_state_exit(spa, NULL, 0);
+		if (suspend) {
+			zio_suspend(spa, NULL, ZIO_SUSPEND_IOERR);
+			zio_resume_wait(spa);
```

Comment:
I am not really familiar with pool suspension, but do we need to block the async thread when it happen?

Reply:
No, spa async will wait for the zios that are in suspended state.  And the spa deadman is ignored for a suspended pool.

Reply:
I am not sure I understand the "No".  To avoid possible misunderstanding, I wanted to say: do we really need this `zio_resume_wait()` call here in `spa_async_thread()`?

Reply:
Ah, I misunderstood the original question.  This is a good question to ask, but maybe not a clear answer is possible. 

If we wait here for the `spa_suspended` state to clear, then any remaining async tasks requested will not proceed until the pool is un-suspended. Waiting here will also hold up any new `spa_async_suspend()` requests, so for example, a `spa_unload()` would have to wait on this as well.  

Also,  `spa_async_dispatch()` won't create a new `spa_async_thread` when there's an active one already, so we could miss out on new async tasks requests that are processed before the `SPA_ASYNC_FAULT_VDEV` is processed.  I think this only means that they won't get processed until the next txg sync occurs. 

Typically, and in the test case that was added here, there are not multiple async tasks being requested at the same time.  So I can't say that it matters either way.  All things being equal, we could go for simplicity and drop the `zio_resume_wait()` here.